modules/metastorage-client/src/main/java/org/apache/ignite/metastorage/client/MetaStorageService.java

```diff
@@ -0,0 +1,334 @@
+     * @see Operation
+     */
+    //TODO: https://issues.apache.org/jira/browse/IGNITE-14269: will be replaced by conditional multi update.
+    @NotNull
+    CompletableFuture<Entry> getAndInvoke(@NotNull byte[] key, @NotNull Condition condition,
+                                          @NotNull Operation success, @NotNull Operation failure);
+
+    /**
+     * Retrieves entries for the given key range in lexicographic order. Entries will be filtered out by upper bound
+     * of given revision number.
+     *
+     * @param keyFrom Start key of range (inclusive). Couldn't be {@code null}.
+     * @param keyTo End key of range (exclusive). Could be {@code null}.
+     * @param revUpperBound  The upper bound for entry revision. {@code -1} means latest revision.
+     * @return Cursor built upon entries corresponding to the given range and revision.
+     * @throws OperationTimeoutException If the operation is timed out.
+     * @throws CompactedException If the desired revisions are removed from the storage due to a compaction.
+     * @see Entry
+     */
+    @NotNull
+    Cursor<Entry> range(@NotNull byte[] keyFrom, @Nullable byte[] keyTo, long revUpperBound);
+
+    /**
+     * Retrieves entries for the given key range in lexicographic order. Short cut for
+     * {@link #range(byte[], byte[], long)} where {@code revUpperBound == -1}.
+     *
+     * @param keyFrom Start key of range (inclusive). Couldn't be {@code null}.
+     * @param keyTo End key of range (exclusive). Could be {@code null}.
+     * @return Cursor built upon entries corresponding to the given range and revision.
+     * @throws OperationTimeoutException If the operation is timed out.
+     * @throws CompactedException If the desired revisions are removed from the storage due to a compaction.
+     * @see Entry
+     */
+    @NotNull
+    Cursor<Entry> range(@NotNull byte[] keyFrom, @Nullable byte[] keyTo);
+
+    /**
+     * Subscribes on meta storage updates matching the parameters.
+     *
+     * @param keyFrom Start key of range (inclusive). Could be {@code null}.
+     * @param keyTo End key of range (exclusive). Could be {@code null}.
+     * @param revision Start revision inclusive. {@code 0} - all revision,
+     * {@code -1} - latest revision (accordingly to current meta storage state).
+     * @param lsnr Listener which will be notified for each update.
+     * @return Subscription identifier. Could be used in {@link #stopWatch} method in order to cancel subscription.
+     * @throws OperationTimeoutException If the operation is timed out. Will be thrown on getting future result.
+     * @throws CompactedException If the desired revisions are removed from the storage due to a compaction.
+     * Will be thrown on getting future result.
+     * @see Entry
+     */
+    @NotNull
+    //TODO: UUID Should be replaced by IgniteUUID when it will be introduced.
+    CompletableFuture<UUID> watch(@Nullable byte[] keyFrom, @Nullable byte[] keyTo,
```

Comment:
There is a time window between the moment this method returns and the moment the result future obtains a watch ID. Within this window it is impossible to cancel the watch and do the cleanup.
Probably, we should generate the watch ID on the client side and return the ID immediately?

Reply:
Or we should properly handle future.cancel() call, but this will be harder to handle from the outside because the external code should either work with future, or with watch ID.

Reply:
@agoncharuk 
> Probably, we should generate the watch ID on the client side and return the ID immediately?

IMHO, we still need return `CompletableFuture` because this operation is remote by nature. We can try to guarantee that the subscription will be destroyed correctly because the `UUID`, in fact, will be generated on the client side. Moreover, where possible exceptions (like OperationTimeoutException) should be handled in case of immediate return?

Reply:
@agoncharuk 

> Or we should properly handle future.cancel() call, but this will be harder to handle from the outside because the external code should either work with future, or with watch ID.

Agree. I think we can postpone final solution to the implementation. Both approaches have disadvantages. In this situation I can't foresee all possible problems.